### DISCUSSION TITLE
Can the error added in #2422 be made opt-in (or less ideally, opt-out)?

### DISCUSSION THREAD

**Discussion by mynewestgitaccount** [encode/starlette#2446] [Potential Issue] [answered]
Recently (in #2422), an error was added to Starlette when Config is used without the `env_file` input leading to a valid file.

Until now, I would use an env file when testing locally and environment variables (without any env file) when deploying in a container. However, this change means that I either need to switch to setting environment variables locally (somewhat tedious) or to creating an env file when deploying.

Would it be possible to make this error opt-in (or even opt-out) to avoid this incompatibility? If changing this is not in the cards (or if the decision is made to make this error opt-out), I would suggest updating the documentation [here](https://www.starlette.io/config/#configuration-precedence), since the new error that gets raised is currently not mentioned and the documentation implies that my previous use would be acceptable.

<hr>

Edit: Looking at the linked commits at the bottom of #2422 (after the merge/release), it seems I am not the only one affected by this.

**Comment by Kludex** [comment]
Can you show me how this part of the code looks like on your project?

**Reply by mynewestgitaccount** [reply]
The core logic looks something like this:

```python
from starlette.config import Config

config = Config(".env")

env1 = config("ENV1")
```

For local development, I would use an env file named `.env` with the variables I needed to avoid needing to manage environment variables.
For containerized deployments, I would define the needed environment variables and would not have the `.env` file.

Now (following #2422), those containerized deployments fail with a `FileNotFoundError`.

Essentially, things have switched from "fail slow, only if lookup fails" (as described in [the configuration precedence section of the documentation](https://www.starlette.io/config/#configuration-precedence)) to "fail fast, if a specified file is missing".

<hr>

I understand the motivation for the change (the author described it in #2101). However, couldn't the same effect be accomplished using a default False (or even default True if needed) `check_file` (arbitrary name for this example) keyword argument to Config like the below?

```diff
class Config:
    def __init__(
        self,
        env_file: typing.Optional[typing.Union[str, Path]] = None,
        environ: typing.Mapping[str, str] = environ,
        env_prefix: str = "",
+       *,
+       check_file: bool = False,
    ) -> None:
        self.environ = environ
        self.env_prefix = env_prefix
        self.file_values: typing.Dict[str, str] = {}
        if env_file is not None:
-           if not os.path.isfile(env_file):
+           if os.path.isfile(env_file):
+               self.file_values = self._read_file(env_file)
+           elif check_file:
                raise FileNotFoundError(f"Config file '{env_file}' not found.")
-           self.file_values = self._read_file(env_file)
```

This would allow the new error to be surfaced for those who want it without requiring it for those who don't.

**Reply by Kludex** [reply]
Can you do the following?
```py
try:
    config = Config(".env")
except FileNotFoundError:
    config = Config()
```

**Reply by mynewestgitaccount** [reply]
I could, but that would require a code change in any application following the previously-working (and documentation-approved) pattern. 🥲

**Comment by rambo** [comment]
I also suddenly have my docker containers failing to start after routine dependency update (obviously they use ENV directly instead of creating a .env inside). Making breaking changes should probably require a bit more discussion on the merits of breaking things that were documented as working.

**Comment by aswanidutt87** [comment]
our unit tests are failing because of this change "FileNotFoundError: Config file '.env' not found."

**Comment by Kludex** [comment, marked as answer]
PR welcome to replace the error by a warning.

**Reply by mathause** [reply]
I opened  #2485 for this. Let me know if you want any changes.

**Reply by rambo** [reply]
Everyone is fast tonigh, also opened https://github.com/encode/starlette/pull/2486 but it's midnight. Anyway I *strongly* urge that the default behaviour is what it used to be as explained in my PR and warnings would be opt-in for those who only ever use config files.

**Comment by isaackogan** [comment]
Same issue, glad it was solved.

**Comment by Kludex** [comment]
Locking this discussion to avoid spam. The "issue" was solved.
